### ISSUE TITLE
[BUG]Azure Linux: "Cannot evict pod as it would violate the pod's disruption budget" during Node Migration

### ISSUE THREAD

**Issue by amitmavgupta** [Azure/AKS#3980]
**Describe the bug**
Azure Linux: "Cannot evict pod as it would violate the pod's disruption budget" during Node Migration
**To Reproduce**

1. As per the migration document https://learn.microsoft.com/en-us/azure/azure-linux/tutorial-azure-linux-migration#2---drain-the-existing-nodes, pdb should be >=1 and that is the case in the current setup we have.

`kubectl get pdb -A
NAMESPACE     NAME                 MIN AVAILABLE   MAX UNAVAILABLE   ALLOWED DISRUPTIONS   AGE
kube-system   coredns-pdb          1               N/A               1                     68m
kube-system   konnectivity-agent   1               N/A               1                     68m
kube-system   metrics-server-pdb   1               N/A               1                     68m`

But when nodes are drained, see these error messages which seem to suggest otherwise. **The nodes are eventually drained** but why are we seeing these error messages?

`kubectl drain aks-nodepool1-39168763-vmss000000 aks-nodepool1-39168763-vmss000001 aks-nodepool1-39168763-vmss000002 --ignore-daemonsets --delete-emptydir-data
node/aks-nodepool1-39168763-vmss000000 already cordoned
node/aks-nodepool1-39168763-vmss000001 already cordoned
node/aks-nodepool1-39168763-vmss000002 already cordoned
Warning: ignoring DaemonSet-managed Pods: cilium-test/host-netns-7l4h2, kube-system/azure-ip-masq-agent-w82p5, kube-system/cilium-s5njz, kube-system/cloud-node-manager-t28np, kube-system/csi-azuredisk-node-2rdfd, kube-system/csi-azurefile-node-5zlsh, kube-system/kube-proxy-b2c7k
evicting pod kube-system/konnectivity-agent-68f47fd869-2gqdq
evicting pod cilium-test/echo-external-node-6b6b6dcc9b-rqbft
evicting pod kube-system/coredns-76b9877f49-gqj4v
pod/konnectivity-agent-68f47fd869-2gqdq evicted
pod/coredns-76b9877f49-gqj4v evicted
pod/echo-external-node-6b6b6dcc9b-rqbft evicted
node/aks-nodepool1-39168763-vmss000000 drained
Warning: ignoring DaemonSet-managed Pods: cilium-test/host-netns-9jvr6, kube-system/azure-ip-masq-agent-vd99d, kube-system/cilium-2qldt, kube-system/cloud-node-manager-gqrgj, kube-system/csi-azuredisk-node-vs2c6, kube-system/csi-azurefile-node-pqvrg, kube-system/kube-proxy-lwmz9
evicting pod kube-system/metrics-server-585676dc6-7vnw9
evicting pod kube-system/konnectivity-agent-68f47fd869-p84xs
evicting pod kube-system/cilium-operator-5bd8c6449f-vpvkr
evicting pod kube-system/coredns-76b9877f49-w8bjf
pod/konnectivity-agent-68f47fd869-p84xs evicted
pod/metrics-server-585676dc6-7vnw9 evicted
pod/cilium-operator-5bd8c6449f-vpvkr evicted
pod/coredns-76b9877f49-w8bjf evicted
node/aks-nodepool1-39168763-vmss000001 drained
Warning: ignoring DaemonSet-managed Pods: cilium-test/host-netns-lf2pl, kube-system/azure-ip-masq-agent-8rm9s, kube-system/cilium-bzh2z, kube-system/cloud-node-manager-6d2rg, kube-system/csi-azuredisk-node-f5sf4, kube-system/csi-azurefile-node-qqzdh, kube-system/kube-proxy-fjrdv
evicting pod kube-system/metrics-server-585676dc6-2ddbr
evicting pod kube-system/coredns-autoscaler-85f7d6b75d-p72nt
evicting pod cilium-test/echo-external-node-6b6b6dcc9b-wrd8n
evicting pod kube-system/cilium-operator-5bd8c6449f-mgtjl
error when evicting pods/"metrics-server-585676dc6-2ddbr" -n "kube-system" (will retry after 5s): Cannot evict pod as it would violate the pod's disruption budget.
pod/coredns-autoscaler-85f7d6b75d-p72nt evicted
pod/cilium-operator-5bd8c6449f-mgtjl evicted
evicting pod kube-system/metrics-server-585676dc6-2ddbr
error when evicting pods/"metrics-server-585676dc6-2ddbr" -n "kube-system" (will retry after 5s): Cannot evict pod as it would violate the pod's disruption budget.
evicting pod kube-system/metrics-server-585676dc6-2ddbr
error when evicting pods/"metrics-server-585676dc6-2ddbr" -n "kube-system" (will retry after 5s): Cannot evict pod as it would violate the pod's disruption budget.
evicting pod kube-system/metrics-server-585676dc6-2ddbr
error when evicting pods/"metrics-server-585676dc6-2ddbr" -n "kube-system" (will retry after 5s): Cannot evict pod as it would violate the pod's disruption budget.
evicting pod kube-system/metrics-server-585676dc6-2ddbr
error when evicting pods/"metrics-server-585676dc6-2ddbr" -n "kube-system" (will retry after 5s): Cannot evict pod as it would violate the pod's disruption budget.
evicting pod kube-system/metrics-server-585676dc6-2ddbr
error when evicting pods/"metrics-server-585676dc6-2ddbr" -n "kube-system" (will retry after 5s): Cannot evict pod as it would violate the pod's disruption budget.
evicting pod kube-system/metrics-server-585676dc6-2ddbr
error when evicting pods/"metrics-server-585676dc6-2ddbr" -n "kube-system" (will retry after 5s): Cannot evict pod as it would violate the pod's disruption budget.
pod/echo-external-node-6b6b6dcc9b-wrd8n evicted
evicting pod kube-system/metrics-server-585676dc6-2ddbr
error when evicting pods/"metrics-server-585676dc6-2ddbr" -n "kube-system" (will retry after 5s): Cannot evict pod as it would violate the pod's disruption budget.
evicting pod kube-system/metrics-server-585676dc6-2ddbr
error when evicting pods/"metrics-server-585676dc6-2ddbr" -n "kube-system" (will retry after 5s): Cannot evict pod as it would violate the pod's disruption budget.
evicting pod kube-system/metrics-server-585676dc6-2ddbr
error when evicting pods/"metrics-server-585676dc6-2ddbr" -n "kube-system" (will retry after 5s): Cannot evict pod as it would violate the pod's disruption budget.
evicting pod kube-system/metrics-server-585676dc6-2ddbr
error when evicting pods/"metrics-server-585676dc6-2ddbr" -n "kube-system" (will retry after 5s): Cannot evict pod as it would violate the pod's disruption budget.
evicting pod kube-system/metrics-server-585676dc6-2ddbr
error when evicting pods/"metrics-server-585676dc6-2ddbr" -n "kube-system" (will retry after 5s): Cannot evict pod as it would violate the pod's disruption budget.
evicting pod kube-system/metrics-server-585676dc6-2ddbr
pod/metrics-server-585676dc6-2ddbr evicted
node/aks-nodepool1-39168763-vmss000002 drained`

**Expected behavior**
Since PDB is =1, why do we see these error messages?

**Screenshots**

**Environment (please complete the following information):**
 -  CLI Version 2.53.1
 - Kubernetes version 1.26.6
 
**Additional context**

@wedaly

**Comment by wedaly** [comment]
I think this is going to be difficult to troubleshoot without more information. Any chance you could provide the exact commands you ran to repro this issue? If not, I'd recommend opening an Azure support ticket with the FQDN of the cluster so support team can investigate.

**Comment by amitmavgupta** [comment]
Hi @wedaly Sure. I have also shared the FQDN offline.

`az group create --name azpcoverlayal --location westus2`

```
az aks create -n azpcoverlayal -g azpcoverlayal -l westus2 \
  --network-plugin azure \
  --network-plugin-mode overlay \
  --pod-cidr 192.168.0.0/16 \
  --network-dataplane cilium \
  --os-sku AzureLinux
``` 

```
kubectl get nodes -o wide
NAME                                 STATUS   ROLES   AGE     VERSION   INTERNAL-IP   EXTERNAL-IP   OS-IMAGE            KERNEL-VERSION     CONTAINER-RUNTIME
aks-alnodepool-36990708-vmss000000   Ready    agent   5m35s   v1.26.6   10.224.0.8    <none>        CBL-Mariner/Linux   5.15.133.1-1.cm2   containerd://1.6.22
aks-alnodepool-36990708-vmss000001   Ready    agent   5m42s   v1.26.6   10.224.0.7    <none>        CBL-Mariner/Linux   5.15.133.1-1.cm2   containerd://1.6.22
aks-nodepool1-31955313-vmss000000    Ready    agent   16m     v1.26.6   10.224.0.4    <none>        CBL-Mariner/Linux   5.15.133.1-1.cm2   containerd://1.6.22
aks-nodepool1-31955313-vmss000001    Ready    agent   16m     v1.26.6   10.224.0.6    <none>        CBL-Mariner/Linux   5.15.133.1-1.cm2   containerd://1.6.22
aks-nodepool1-31955313-vmss000002    Ready    agent   16m     v1.26.6   10.224.0.5    <none>        CBL-Mariner/Linux   5.15.133.1-1.cm2   containerd://1.6.22
```

```
kubectl cordon aks-nodepool1-31955313-vmss000000 aks-nodepool1-31955313-vmss000001 aks-nodepool1-31955313-vmss000002
node/aks-nodepool1-31955313-vmss000000 cordoned
node/aks-nodepool1-31955313-vmss000001 cordoned
node/aks-nodepool1-31955313-vmss000002 cordoned
```

```

kubectl get pdb -A
NAMESPACE     NAME                 MIN AVAILABLE   MAX UNAVAILABLE   ALLOWED DISRUPTIONS   AGE
kube-system   coredns-pdb          1               N/A               1                     16m
kube-system   konnectivity-agent   1               N/A               1                     16m
kube-system   metrics-server-pdb   1               N/A               1                     16m

```

```
kubectl drain aks-nodepool1-31955313-vmss000000 aks-nodepool1-31955313-vmss000001 aks-nodepool1-31955313-vmss000002 --ignore-daemonsets --delete-emptydir-data
node/aks-nodepool1-31955313-vmss000000 already cordoned
node/aks-nodepool1-31955313-vmss000001 already cordoned
node/aks-nodepool1-31955313-vmss000002 already cordoned
Warning: ignoring DaemonSet-managed Pods: kube-system/azure-cns-fn745, kube-system/azure-ip-masq-agent-rcrhx, kube-system/cilium-lcp27, kube-system/cloud-node-manager-7q6z8, kube-system/csi-azuredisk-node-hrpkq, kube-system/csi-azurefile-node-x6hbw
evicting pod kube-system/coredns-76b9877f49-tkqqd
evicting pod kube-system/cilium-operator-8cff7865b-72vst
pod/cilium-operator-8cff7865b-72vst evicted
pod/coredns-76b9877f49-tkqqd evicted
node/aks-nodepool1-31955313-vmss000000 drained
Warning: ignoring DaemonSet-managed Pods: kube-system/azure-cns-pvqr4, kube-system/azure-ip-masq-agent-xq4kf, kube-system/cilium-cc42q, kube-system/cloud-node-manager-w6bn6, kube-system/csi-azuredisk-node-f97bb, kube-system/csi-azurefile-node-kxgqx
evicting pod kube-system/metrics-server-585676dc6-zd8c5
evicting pod kube-system/coredns-autoscaler-85f7d6b75d-fhtgz
evicting pod kube-system/coredns-76b9877f49-4knhw
evicting pod kube-system/konnectivity-agent-68cfb6ff54-szz7f
evicting pod kube-system/konnectivity-agent-68cfb6ff54-44fxd
**error when evicting pods/"konnectivity-agent-68cfb6ff54-szz7f" -n "kube-system" (will retry after 5s): Cannot evict pod as it would violate the pod's disruption budget.**
pod/coredns-autoscaler-85f7d6b75d-fhtgz evicted
pod/konnectivity-agent-68cfb6ff54-44fxd evicted
pod/metrics-server-585676dc6-zd8c5 evicted
evicting pod kube-system/konnectivity-agent-68cfb6ff54-szz7f
pod/coredns-76b9877f49-4knhw evicted
pod/konnectivity-agent-68cfb6ff54-szz7f evicted
node/aks-nodepool1-31955313-vmss000001 drained
Warning: ignoring DaemonSet-managed Pods: kube-system/azure-cns-gqts9, kube-system/azure-ip-masq-agent-hh25f, kube-system/cilium-569gt, kube-system/cloud-node-manager-ddhjl, kube-system/csi-azuredisk-node-c4bmq, kube-system/csi-azurefile-node-h6sjv
evicting pod kube-system/cilium-operator-8cff7865b-qxg8f
pod/cilium-operator-8cff7865b-qxg8f evicted
node/aks-nodepool1-31955313-vmss000002 drained
```

**Comment by wedaly** [comment]
thanks for the detailed steps to reproduce! I was able to run those steps, and I'm seeing the same error logs.

Here's what I see:
1. Nodes start draining, removing pods that have PDBs (konnectivity-agent, coredns, and metrics-server).
2. As pods are removed, the deployments schedule new pods to replace them.
3. After pods are scheduled to new nodes, it can take a few seconds before they reach "Running" state. During this time, PDBs could prevent some nodes from draining pods until the replacement pods are ready.
4. Eventually, the replacement pods reach Running state, and the nodes are able to finish draining successfully.

In short, I believe the behavior you reported is expected, so I'm going to close this bug. Thanks!
